src/traits.jl

```diff
@@ -0,0 +1,48 @@
+"""
+    type Cardinality
+
+A trait describing the cardinality of a [`Transform`]. Available cardinalities are:
+[`OneToOne`](@ref), [`ManyToOne`](@ref), [`OneToMany`](@ref), and [`ManyToMany`](@ref).
+"""
+abstract type Cardinality end
+
+"""
+    OneToOne <: Cardinality
+
+Transforms that map each input to exactly one output: `x → y`.
+Examples: [`Power`](@ref), [`Periodic`](@ref).
+"""
+struct OneToOne <: Cardinality end
+
+"""
+    ManyToOne <: Cardinality
+
+Transforms that map many inputs to one output: `(x_1, x_2, ..., x_n) → y`.
+These are typically reduction operations.
+Examples: [`LinearCombination`](@ref).
+"""
+struct ManyToOne <: Cardinality end
+
+"""
+    OneToMany <: Cardinality
+
+Transforms that map one input to many outputs: `x → (y_1, y_2, ..., y_n)`.
+Examples: [`OneHotEncoding`](@ref).
+"""
+struct OneToMany <: Cardinality end
+
+"""
+    ManyToMany <: Cardinality
+
+Transforms that map many inputs to many outputs: `(x_1, x_2, ..., x_m) → (y_1, y_2, ..., y_n)`.
+Examples: Principle Component Analysis (not implemented).
+"""
+struct ManyToMany <: Cardinality end
+
+
+"""
+    cardinality(transform) -> Cardinality
+
+Returns the [`Cardinality`](@ref) of the `transform`.
+"""
+function cardinality end
```

Comment:
I initially thought defining `cardinality` on instances rather than types seemed redundant in a way. And that `cardinality` for each subtype of `AbstractScaling` seemed redundant. But I realised it's how traits are supposed to work right?

Reply:
I'm not sure I understand. 
In this instance I could/should have defined the `cardinality` on `AbstractScaling`, which would have covered `MeanStdScaling` and `IdentityScaling`.

Reply:
Yes that's partly what I was thinking. For some reason thought defining `cardinality` on `AbstractScaling` wouldn't work because it was an abstract type, so I didn't comment on that directly. 

The other point I was making is, why should every instance of a given `Transform` subtype have this trait, when the trait is the same between all instances? Conceptually, cardinality belongs more with the type than the instances.

I backed off on that because I figured traits, as a design pattern, apply to instances. But I'm not even sure about that - is it a valid point?

Reply:
oh right, well they can apply to both. 

You'll see I've changed the code so that cardinality now applies to `AbstractScaling`.
In that case it made more sense to attach it to the type because all concrete subtypes will have the same behaviour.

But I attached it directly to the concrete types of all other transforms because they don't have abstract supertype (besides Transform) to define it for. And if I defined it for `Transform` outright then users won't know they have to define it themselves.

Reply:
Yep that looks fine. Still not sure if I got this across though:

> why should every instance of a given Transform subtype have this trait, when the trait is the same between all instances? Conceptually, cardinality belongs more with the type than the instances.

This means e.g.

```
julia> cardinality(T) where T <: Power = OneToOne()

julia> cardinality(Power) 
OneToOne
```

vs.

```
julia> cardinality(::Power) = OneToOne()

julia> cardinality(Power(2))
OneToOne
```